mountinfo/mountinfo_linux.go

```diff
@@ -128,7 +131,19 @@ func GetMountsFromReader(r io.Reader, filter FilterFunc) ([]*Info, error) {
 }
 
 func parseMountTable(filter FilterFunc) ([]*Info, error) {
-	f, err := os.Open("/proc/self/mountinfo")
+	// We need to lock ourselves to the current OS thread in order to make sure
+	// that the thread referenced by /proc/thread-self stays alive until we
+	// finish parsing the file.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	f, err := os.Open("/proc/thread-self/mountinfo")
+	if os.IsNotExist(err) {
+		// On pre-3.17 kernels (such as CentOS 7), we don't have
```

Comment:
The only thing we could add here is a mechanism to remember whether we have the `/proc/thread-self/mountinfo` and skip trying to open it later if we don't have it.

Not sure if this optimization is worth the trouble -- probably not.

Reply:
I have that optimisation in the runc code, I can copy it here if you like?

Reply:
Yes please. I guess it makes sense (e.g. for kubelet, which is a long lived daemon).